### PR TITLE
Fix "to-many key not allowed here" when using `memberName` filter

### PR DESCRIPTION
**Pull request by ipavlidakis**
### 🔗 Issue Links

Resolves https://github.com/GetStream/ios-issues-tracking/issues/416

### 🎯 Goal

Fix a an error when using the `memberName` filter with ChannelList.

### 🛠 Implementation

Introduced a way to explicitly mark filter keypaths as collections, to enable us to adapt the NSPredicate translation accordingly.

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [x] This change should be manually QAed
- [x] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎁 Meme

image

### CODE REVIEW

File: Sources/StreamChat/Query/Filter.swift [hunk 1]
```diff
@@ -139,14 +143,16 @@ public struct Filter<Scope: FilterScope> {
     public init(
         operator: String,
         key: String?,
-        value: FilterValue
+        value: FilterValue,
+        isACollectionFilter: Bool
```
Comment (nuno-vieira) [inline Sources/StreamChat/Query/Filter.swift:147]: This is probably enough:
```suggestion
        isCollectionFilter: Bool
```

File: Sources/StreamChat/Query/Filter.swift [hunk 2]
```diff
@@ -108,19 +108,23 @@ public struct Filter<Scope: FilterScope> {
     /// a local filtering.
     let keyPathString: String?
 
+    let isACollectionFilter: Bool
```
Comment (nuno-vieira) [inline Sources/StreamChat/Query/Filter.swift:111]:
```suggestion
    /// Whether the `keyPathString` represents an array in the DTO entities.
    let isCollectionFilter: Bool
```

File: Sources/StreamChat/Query/Filter+ChatChannel.swift [hunk 3]
```diff
@@ -180,15 +180,15 @@ extension Filter where Scope == ChannelListFilterScope {
...
+                format: "%K BEGINSWITH[c] %@".prepend(("ANY "), ifCondition: isACollectionFilter),
                 argumentArray: [keyPathString, prefix]
             )
         case .contains where mappedValue is String:
             guard let needle = mappedValue as? String else {
                 return nil
             }
             return NSPredicate(
-                format: "%K CONTAINS %@",
+                format: "%K CONTAINS %@".prepend(("ANY "), ifCondition: isACollectionFilter),
```
Comment (nuno-vieira) [inline Sources/StreamChat/Query/Filter+ChatChannel.swift:191]: Are we sure we need this for contains? 🤔 I don't think so. Can we unit test this?

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,9 +2,11 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 # Upcoming
-
 ### 🔄 Changed
 
+### 🐞 Fixed
```
Comment (nuno-vieira) [inline CHANGELOG.md:7]:
```suggestion
## StreamChat
### 🐞 Fixed
```

File: Sources/StreamChat/Query/Filter+ChatChannel.swift (hunk 3 above)
Comment (ipavlidakis) [reply Sources/StreamChat/Query/Filter+ChatChannel.swift]: This is actually needed for the cases like this one `.contains(.memberName, value: "test")`. I added the related testCase.

File: Tests/StreamChatTests/Controllers/ChannelListController/ChannelListController_Tests.swift [hunk 5]
```diff
@@ -1619,6 +1619,63 @@ final class ChannelListController_Tests: XCTestCase {
...
+
+        try assertFilterPredicate(
+            .autocomplete(.name, text: "test"),
+            channelsInDB: [
+                .dummy(channel: .dummy(cid: .unique, name: "channelA")),
+                .dummy(channel: .dummy(cid: .unique, name: "channelB")),
+                .dummy(channel: .dummy(cid: cid, name: "test"))
+            ],
+            expectedResult: [cid]
+        )
```
Comment (nuno-vieira) [inline Tests/StreamChatTests/Controllers/ChannelListController/ChannelListController_Tests.swift:1654]: In this test we are not testing the "autocomplete" feature, we need to verify if indeed it returns the channel with the correct "prefix"
```suggestion
        let cid = ChannelId.unique

        try assertFilterPredicate(
            .autocomplete(.name, text: "test"),
            channelsInDB: [
                .dummy(channel: .dummy(cid: .unique, name: "channelA")),
                .dummy(channel: .dummy(cid: .unique, name: "channelB")),
                .dummy(channel: .dummy(cid: cid, name: "testChannel"))
            ],
            expectedResult: [cid]
        )
```

File: Tests/StreamChatTests/Controllers/ChannelListController/ChannelListController_Tests.swift [hunk 6]
```diff
@@ -1619,6 +1619,63 @@ final class ChannelListController_Tests: XCTestCase {
...
+                ])),
+                .dummy(channel: .dummy(cid: .unique, members: [
+                    .dummy(user: .dummy(userId: .unique, name: "userB"))
+                ])),
+                .dummy(channel: .dummy(cid: cid, members: [
+                    .dummy(user: .dummy(userId: .unique, name: "test"))
+                ]))
+            ],
+            expectedResult: [cid]
+        )
```
Comment (nuno-vieira) [inline Tests/StreamChatTests/Controllers/ChannelListController/ChannelListController_Tests.swift:1640]: Same here:
```suggestion
        let cid = ChannelId.unique

        try assertFilterPredicate(
            .autocomplete(.memberName, text: "test"),
            channelsInDB: [
                .dummy(channel: .dummy(cid: .unique, members: [
                    .dummy(user: .dummy(userId: .unique, name: "userA")),
                    .dummy(user: .dummy(userId: .unique, name: "userC"))
                ])),
                .dummy(channel: .dummy(cid: .unique, members: [
                    .dummy(user: .dummy(userId: .unique, name: "userB"))
                ])),
                .dummy(channel: .dummy(cid: cid, members: [
                    .dummy(user: .dummy(userId: .unique, name: "testChannel"))
                ]))
            ],
            expectedResult: [cid]
        )
```

**Review by nuno-vieira — APPROVED** [review]
LGTM! ✅

**Pull request merged**